### PR TITLE
RFD 156 - Kubernetes Replica Count Management Server Source Code

### PR DESCRIPTION
**Pull request by mrkooll**
This pull request contains the implementation of our Kubernetes Replica Count Management Server.

### CODE REVIEW

File: sre-server/main.go [hunk 1]
```diff
@@ -42,23 +103,218 @@ func run(args []string) error {
...
+}
+
+// deploymentHandler is an HTTP handler for the deployment API.
+type deploymentHandler struct {
+	clientset *kubernetes.Clientset
+	lister    listersv1.DeploymentLister
+}
+
+func writeMessage(w http.ResponseWriter, statusCode int, data any) {
+	d, _ := json.Marshal(data)
```
Comment (sclevine) [inline sre-server/main.go]: Recommend checking this error instead of assuming `d` is okay to return in case of error.

File: sre-server/main.go [hunk 2]
```diff
@@ -42,23 +103,218 @@ func run(args []string) error {
...
+
+		if spec.ReplicaCount < 0 {
+			writeError(w, http.StatusBadRequest, "missing or invalid replicaCount")
+			return
+		}
+
+		// fetch the deployment
+		deploymentClient := h.clientset.AppsV1().Deployments(namespace)
+
+		deployment, err := deploymentClient.Get(r.Context(), name, metav1.GetOptions{})
```
Comment (sclevine) [inline sre-server/main.go]: Why not use the deployment stored in the informer?

File: sre-server/main.go [hunk 3]
```diff
@@ -42,23 +103,218 @@ func run(args []string) error {
...
+		if err != nil {
+			writeError(w, http.StatusInternalServerError, "error while fetching deployment: "+err.Error())
+			return
+		}
+		// set replica count
+		deployment.Spec.Replicas = &spec.ReplicaCount
+		// update the deployment
+		_, err = deploymentClient.Update(r.Context(), deployment, metav1.UpdateOptions{})
```
Comment (sclevine) [inline sre-server/main.go]: This will fail due to conflict if unrelated fields change between Get and Put.

File: sre-server/main.go [hunk 4]
```diff
@@ -42,23 +103,218 @@ func run(args []string) error {
...
 }
 
 // ServeHTTP implements http.Handler
 func (h *healthzHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	pods, err := h.clientset.CoreV1().Pods("").List(r.Context(), metav1.ListOptions{})
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusServiceUnavailable)
-		return
+	if h.informer.HasSynced() {
+		_, err := h.clientset.CoreV1().Pods("").List(r.Context(), metav1.ListOptions{Limit: 1})
```
Comment (sclevine) [inline sre-server/main.go]: Should this assume the client has permission to list pods? Is there a cheaper readiness check we could use to verify cluster connectivity?

File: sre-server/main.go [hunk 5]
```diff
@@ -42,23 +103,218 @@ func run(args []string) error {
...
+		// Handle PUT request
+		// request body spec
+		type ReplicaCountSpec struct {
+			ReplicaCount int32 `json:"replicaCount"`
+		}
+
+		// get replica count
+		// var spec ReplicaCountSpec
+		spec := ReplicaCountSpec{ReplicaCount: -1}
+		err := json.NewDecoder(r.Body).Decode(&spec)
```
Comment (sclevine) [inline sre-server/main.go]: Is it safe to read the entire request body here?

File: sre-server/main.go [hunk 6]
```diff
@@ -42,23 +103,218 @@ func run(args []string) error {
...
+}
+
+// ServeHTTP implements http.Handler
+func (h *deploymentHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	namespace := vars["namespace"]
+	name := vars["name"]
+
+	// check is deployment exist
+	deploymentSpec, err := h.lister.Deployments(namespace).Get(name)
```
Comment (sclevine) [inline sre-server/main.go]:
```suggestion
	deployment, err := h.lister.Deployments(namespace).Get(name)
```
Technically this has both spec and status.

File: sre-server/main.go [hunk 7]
```diff
@@ -42,23 +103,218 @@ func run(args []string) error {
...
 	// add handlers
-	http.Handle("/healthz", &healthzHandler{clientset: clientset})
+	router.Handle("/api/v1/namespaces/{namespace}/deployments/{name}/replicas",
+		&deploymentHandler{clientset: clientset, lister: deploymentLister})
+	router.Handle("/api/v1/deployments", &deploymentsHandler{lister: deploymentLister})
+	router.Handle("/api/v1/healthz", &healthzHandler{clientset: clientset, informer: deploymentInformer})
+	router.Handle("/api/v1/pingz", &pingzHandler{})
+
+	// Create HTTP Server with TLS Config
+	server := &http.Server{
```
Comment (sclevine) [inline sre-server/main.go:139]: How do I upgrade the service without dropping requests?

File: sre-server/main.go (hunk 5 above)
Comment (mrkooll) [reply sre-server/main.go]: It make sense to limit a size of parsed request body. Will do

File: sre-server/main.go (hunk 1 above)
Comment (mrkooll) [reply sre-server/main.go]: Will do.

File: sre-server/main.go (hunk 2 above)
Comment (mrkooll) [reply sre-server/main.go]: My bad. Will use informer's deployment here.

File: sre-server/main.go (hunk 3 above)
Comment (mrkooll) [reply sre-server/main.go]: Will add retry logic.

File: sre-server/main.go (hunk 4 above)
Comment (mrkooll) [reply sre-server/main.go]: Yes. It's stupid. Will use the different api check

File: sre-server/main.go (hunk 3 above)
Comment (sclevine) [reply sre-server/main.go]: That seems complex. Do we need to make an update request with optimistic locking on the entire spec? Are conflicts useful for this use case?

File: sre-server/main.go (hunk 7 above)
Comment (mrkooll) [reply sre-server/main.go:139]: Will add graceful shutdown for it.

**Comment by mrkooll** [comment]
Updated according to the review's comments.

File: sre-server/main.go (hunk 3 above)
Comment (mrkooll) [reply sre-server/main.go]: Do you mean just use patch instead of using deployment update?

Will rewrite to use patch - it should be simplest solution IMHO.

**Comment by mrkooll** [comment]
Updated to use deployment patching instead of updating to minimize possibility of conflicts.

File: sre-server/main.go (hunk 6 above)
Comment (mrkooll) [reply sre-server/main.go]: done

File: sre-server/main.go (hunk 3 above)
Comment (sclevine) [reply sre-server/main.go]: Looks good! Could alternatively use the `/scale` subresource to cut the deployment spec out entirely 🙂

**Review by sclevine — APPROVED** [review]

**Review by jimbishopp — APPROVED** [review]

**Comment by jimbishopp** [comment]
@mrkooll Thanks for this! You now have two approvals so feel free to merge this and open the final automation PR.
